alf/algorithms/merlin_algorithm.py

```diff
@@ -0,0 +1,441 @@
+        ]
+        loss = tf.add_n([decoder_loss.loss for decoder_loss in decoder_losses])
+        decoder_losses = tf.nest.pack_sequence_as(self._decoders,
+                                                  decoder_losses)
+        return LossInfo(loss=loss, extra=decoder_losses)
+
+    def train_step(self, inputs, state: MBPState):
+        """Train one step.
+
+        Args:
+            inputs (tuple): a tuple of (observation, action)
+        """
+        observation, _ = inputs
+        latent_vector, kld, next_state = self.encode_step(inputs, state)
+
+        # TODO: decoder for action
+        decoder_loss = self.decode_step(latent_vector, observation)
+
+        return AlgorithmStep(
+            outputs=latent_vector,
+            state=next_state,
+            info=LossInfo(
+                loss=decoder_loss.loss + kld,
+                extra=MBPLossInfo(decoder=decoder_loss, vae=kld)))
+
+
+class MemoryBasedActor(OnPolicyAlgorithm):
+    """The policy module for MERLIN model."""
+
+    def __init__(
+            self,
+            action_spec,
+            memory: MemoryWithUsage,
+            num_read_keys=1,
+            lstm_size=(100, 100),  # not sure what Merlin uses
+            latent_dim=20,
+            loss=None,
+            name="mba"):
+        """Create the policy module of MERLIN.
+
+        Args:
+            action_spec (nested BoundedTensorSpec): representing the actions.
+            memory (MemoryWithUsage): the memory module from MemoryBasedPredictor
+            num_read_keys (int): number of keys for reading memory.
+            latent_dim (int): the dimension of the hidden represenation of VAE.
+            lstm_size (list[int]): size of lstm layers
+            loss (None|ActorCriticLoss): an object for calculating the loss
+                for reinforcement learning. If None, a default ActorCriticLoss
+                will be used.
+            name (str): name of the algorithm.
+        """
+        # This is different from Merlin LSTM. This rnn only uses the output
+        # from ths last LSTM layer, while Merlin uses outputs from all LSTM
+        # layers
+        rnn = make_lstm_cell(lstm_size, name=name + "/lstm")
+
+        actor_input_dim = (
+            latent_dim + lstm_size[-1] + num_read_keys * memory.dim)
+
+        actor_net = ActorDistributionNetwork(
+            input_tensor_spec=TensorSpec((actor_input_dim, ),
+                                         dtype=tf.float32),
+            output_tensor_spec=action_spec,
+            fc_layer_params=(200, ),
+            activation_fn=tf.keras.activations.tanh,
+            name=name + "/actor_net")
+
+        super(MemoryBasedActor, self).__init__(
+            action_spec=action_spec,
+            train_state_spec=get_rnn_cell_state_spec(rnn),
+            action_distribution_spec=actor_net.output_spec)
+
+        self._loss = ActorCriticLoss(action_spec) if loss is None else loss
```

Comment:
make ActorCriticLoss gin.configurable so it could let client enable more powerful features like td/gae

Reply:
It's already gin.configurable